packages/commonwealth/server/util/cosmosProxy.ts

```diff
@@ -27,36 +28,78 @@ function setupCosmosProxy(
     bodyParser.text(),
     calcCosmosRPCCacheKeyDuration,
     cacheDecorator.cacheMiddleware(
-      defaultCacheDuration,
+      DEFAULT_CACHE_DURATION,
       lookupKeyDurationInReq,
     ),
     async function cosmosProxy(req, res) {
       log.trace(`Got request: ${JSON.stringify(req.body, null, 2)}`);
+      let cosmos_chain_id, chainNodeUrl, previouslyFailed;
       try {
-        log.trace(`Querying cosmos endpoint for chain: ${req.params.chain}`);
+        const chainParam = req.params.chain;
         const chain = await models.Community.findOne({
-          where: { id: req.params.chain },
+          where: { id: chainParam },
           include: models.ChainNode,
         });
         if (!chain) {
-          throw new AppError('Invalid chain');
+          throw new AppError(`Invalid chain: ${chainParam}`);
         }
-        log.trace(`Found cosmos endpoint: ${chain.ChainNode.url}`);
-        const response = await axios.post(chain.ChainNode.url, req.body, {
-          headers: {
-            origin: 'https://commonwealth.im',
-          },
-        });
+
+        let response;
+        const chainNode = chain.ChainNode;
+        chainNodeUrl = chainNode?.url?.trim();
+        let useProxy = !chainNodeUrl;
+        cosmos_chain_id = chainNode?.cosmos_chain_id;
+        previouslyFailed = chainNode?.health === NodeHealth.Failed;
+        if (previouslyFailed) {
+          const lastUpdate = chainNode?.updated_at?.getTime();
+          const healthPauseTimeout = new Date(
+            lastUpdate + FALLBACK_NODE_DURATION,
+          );
+          useProxy = previouslyFailed && new Date() > healthPauseTimeout;
+        }
+
+        if (chainNodeUrl && !useProxy) {
+          log.trace(`Found cosmos endpoint: ${chainNodeUrl}`);
+          response = await axios
+            .post(chainNodeUrl, req.body, {
+              headers: {
+                origin: 'https://commonwealth.im',
+              },
+            })
+            .catch(async (err) => {
+              log.trace(`Error: ${err.message}`);
+              await flagFailedNode(
+                previouslyFailed,
+                cosmos_chain_id,
+                chainNodeUrl,
+              );
+              return queryRPCProxy(req, chain);
+            });
+
+          await flagHealthyNode(
+            response,
+            previouslyFailed,
+            cosmos_chain_id,
+            chainNodeUrl,
+          );
+        } else {
+          response = await queryRPCProxy(req, chain.ChainNode.cosmos_chain_id);
+        }
+
         log.trace(
           `Got response from endpoint: ${JSON.stringify(
             response.data,
             null,
             2,
           )}`,
         );
+
         return res.send(response.data);
       } catch (err) {
-        res.status(500).json({ message: err.message });
+        await flagFailedNode(previouslyFailed, cosmos_chain_id, chainNodeUrl);
```

Comment:
Basically we need cosmos_chain_id to be set, otherwise this function would throw an error itself. Therefore I think it would be good to have a more fine grained try-catch, instead of try-catching the whole cosmosProxy. Can we narrow down the try catch block to the only call that could error that we haven't caught?

So maybe around this function call?
```
response = await queryRPCProxy(req, chain.ChainNode.cosmos_chain_id);
```

Reply:
> Can we narrow down the try catch block to the only call that could error that we haven't caught?

I agree with this, and I thought about doing this.

But in practice it has been difficult to anticipate why a node fails. It has been a number of different reasons. With this strategy, we temporarily use the fallback if the node fails for any reason. And then after 5 minutes, we go back to trying the preferred node.

Reply:
>Basically we need cosmos_chain_id to be set, otherwise this function would throw an error itself.

will handle this